File: models/Heart.ts
import Position from "../interfaces/Position";
import Damageable from "../interfaces/Damageable";
import Vehicle from './Vehicle';
import DamageRemover from "../interfaces/DamageRemover";

class Heart extends DamageRemover {
  private x: number = 0;
  private y: number = 0;

  static RADIUS = 2;

  constructor (
    x: number = 0,
    y: number = 0,
  ) {
    super()
    this.x = x;
    this.y = y;
  }

  getPosition(): Position {
    return { x: this.x, y: this.y };
  }

  getBoundaries(): Array<Position> {
    return [
      { x: this.x, y: this.y },
      { x: this.x + 3, y: this.y - 3 },
      { x: this.x + 6, y: this.y },
      { x: this.x, y: this.y + 6 },
      { x: this.x - 6, y: this.y },
      { x: this.x - 3, y: this.y - 3 },
    ];
  }

  getDamageToRemove(damageable: Damageable): number {
    if (damageable instanceof Vehicle) {
      return 30
    }
    return 2
  }
  
}

export default Heart;

File: rust/models/Heart.rs
use crate::interfaces::{Position, Damageable, Damager};

pub struct Heart {
    x: f64,
    y: f64,
}

impl Heart {
    pub const RADIUS: f64 = 2.0;

    pub fn new(x: f64, y: f64) -> Self {
        Heart { x, y }
    }

    pub fn get_position(&self) -> Position {
        Position { x: self.x, y: self.y }
    }

    pub fn get_boundaries(&self) -> Vec<Position> {
        vec![
            Position { x: self.x, y: self.y },
            Position { x: self.x + 3.0, y: self.y - 3.0 },
            Position { x: self.x + 6.0, y: self.y },
            Position { x: self.x, y: self.y + 6.0 },
            Position { x: self.x - 6.0, y: self.y },
            Position { x: self.x - 3.0, y: self.y - 3.0 },
        ]
    }

    pub fn get_damage_to_remove(&self, damageable: &Damageable) -> f64 {
        if let Some(vehicle) = damageable.as_vehicle() {
            return 30.0;
        }
        2.0
    }
}
